lean4/src/putnam_2023_a6.lean

```diff
@@ -0,0 +1,22 @@
+import Mathlib
+
+open Finset
+
+abbrev putnam_2023_a6_solution : Set ℕ := sorry
+-- {n : ℕ | 0 < n}
+/--
+Alice and Bob play a game in which they take turns choosing integers from $1$ to $n$. Before any integers are chosen, Bob selects a goal of 'odd' or 'even'. On the first turn, Alice chooses one of the $n$ integers. On the second turn, Bob chooses one of the remaining integers. They continue alternately choosing one of the integers that has not yet been chosen, until the $n$th turn, which is forced and ends the game. Bob wins if the parity of $\{k : \mbox{the number $k$ was chosen on the $k$th turn}\}$ matches his goal. For which values of $n$ does Bob have a winning strategy?
+-/
+theorem putnam_2023_a6
+    (IsValidGame : List ℕ → Prop)
+    (IsValidGame_def : ∀ g, IsValidGame g ↔ g.Nodup ∧ (∀ i ∈ g, i ∈ Icc 1 g.length))
+    (parityOf : List ℕ → ZMod 2)
+    (parityOf_def : ∀ g, parityOf g = ((range g.length).filter fun i ↦ g[i]! = i + 1).card)
```

Comment:
Q: Is it a standard convention to use coercions implicitly from a "larger" type to a "smaller" type? Usually we have used coercions for the other direction. I am curious if this is normal in Mathlib - this might hurt readability slightly.

Reply:
I'm afraid I don't understand this question, can you point me at an example where you have used a coercion in the "other direction"?

Reply:
Ah by this I mean a coercion `\N \to \Z`, for example. Here "smaller" and "larger" are a reference to `\N \subset \Z`. This one _seems_ more natural to me, and I might expect a reader to be able to parse this example easier than a coercion "in the other direction." But this is probably negligible, if not an unreasonable take due to inexperience on my part.

Reply:
The job of a coercion is to remove the obligation for us to avoid being explicit about maps for which there is only one possible candidate. Or maybe more precisely, the only possible map that the author could have intended without explicitly saying something else.

The map `ℕ → ℤ` which you highlight is certainly of this nature, and I think the map `ℕ → ZMod 2` is too. In any case, Mathlib has already made this call so I suggest we fall in line.

If you care to see the details, the Infoview tells me that the map is provided by the `NatCast` typeclass and so I ran
```lean
#synth NatCast (ZMod 2)
```
This succeeds with output `Semiring.toNatCast` and after chasing through the class hierarchy a bit I find the map arises because `AddMonoidWithOne` extends `NatCast`.